### DISCUSSION TITLE
openssl 3.0.2

### DISCUSSION THREAD

**Discussion by jfagf** [resiprocate/resiprocate#242] [General]
We need the resiprocate support for openssl 3.0.2 because we want use  
the yocto Long Term Support release 4.0 (kirkstone) with openssl 3.0.2

Is there any intention to implement it in the near future?

Now threre are compiler errors like that:

ssl/OpenSSLInit.cxx: In constructor 'resip::OpenSSLInit::OpenSSLInit()':
| ssl/OpenSSLInit.cxx:76:9: error: 'CRYPTO_set_mem_debug' was not declared in this scope; did you mean 'CRYPTO_set_ex_data'?
|    76 |         CRYPTO_set_mem_debug(1);
|       |         ^~~~~~~~~~~~~~~~~~~~
|       |         CRYPTO_set_ex_data
| ssl/OpenSSLInit.cxx:79:25: error: 'CRYPTO_MEM_CHECK_ON' was not declared in this scope
|    79 |         CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
|       |                         ^~~~~~~~~~~~~~~~~~~
| ssl/OpenSSLInit.cxx:79:9: error: 'CRYPTO_mem_ctrl' was not declared in this scope; did you mean 'CRYPTO_memcmp'?
|    79 |         CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
|       |         ^~~~~~~~~~~~~~~
|       |         CRYPTO_memcmp

Thanks

**Comment by sgodin** [comment]
We don't have a fixed schedule for updating to the latest openssl.  It is
typically done when a core developer needs it for one of their projects.
We would gladly accept pull requests as well.   We use #ifdefs to ensure
older versions still compile.

Scott

**Comment by gjasny** [comment]
You could add a `OPENSSL_API_COMPAT=10101` C(XX)FLAG and the following:
```diff
--- a/rutil/ssl/OpenSSLInit.cxx
+++ b/rutil/ssl/OpenSSLInit.cxx
@@ -72,11 +72,13 @@ OpenSSLInit::OpenSSLInit()
 #if (OPENSSL_VERSION_NUMBER < 0x10100000L) || defined(LIBRESSL_VERSION_NUMBER)
        CRYPTO_malloc_debug_init();
        CRYPTO_set_mem_debug_options(V_CRYPTO_MDEBUG_ALL);
-#else
+#elif !defined(OPENSSL_NO_CRYPTO_MDEBUG)
        CRYPTO_set_mem_debug(1);
 #endif
 
+#ifndef OPENSSL_NO_CRYPTO_MDEBUG
        CRYPTO_mem_ctrl(CRYPTO_MEM_CHECK_ON);
+#endif
 
        SSL_library_init();
        SSL_load_error_strings();
```

Generally I find it hostile of reSIProcate to touch the OpenSSL global state and unconditionally enable the debugging options. We once had a severe regression in network communication throughput which was caused by a completely different component in the process started using resiprocate.

Most of the OpenSSLInit functions are not needed with 1.1.1 and later.

**Reply by jfagf** [reply]
Many thanks for your help.
It goes one step further.
But now there is an error compiling with openssl 3.0.0 in

Security.cxx:959:48: error: 'ERR_GET_FUNC'
OpenSSL changelog says:

https://www.openssl.org/news/changelog.html

The ERR_GET_FUNC() function was removed. With the loss of meaningful function codes, this function can only cause problems for calling applications.

Dow you know how can we avoid this error?

**Reply by gjasny** [reply]
```diff
diff --git a/resip/stack/ssl/Security.cxx b/resip/stack/ssl/Security.cxx
index 8e6f60ec1..e1315bd75 100644
--- a/resip/stack/ssl/Security.cxx
+++ b/resip/stack/ssl/Security.cxx
@@ -962,11 +962,13 @@ BaseSecurity::addPrivateKeyPEM( PEMType type,
          char buffer[120];
          unsigned long err = ERR_get_error();
          ERR_error_string(err, buffer);
+#ifdef ERR_GET_FUNC
          if(ERR_GET_LIB(err) == ERR_LIB_EVP && ERR_GET_FUNC(err) == EVP_F_EVP_DECRYPTFINAL_EX && ERR_GET_REASON(err) == EVP_R_BAD_DECRYPT)
          {
             ErrLog(<< "Could not read private key (error=" << buffer << ") - likely incorrect password provided, may load correctly when transports are added with ap
propriate password");
          }
          else
+#endif
          {
             ErrLog(<< "Could not read private key (error=" << buffer << ") from <" << privateKeyPEM << ">" );
          }
```

**Reply by Kingcompton2** [reply]
I need help

**Comment by jfagf** [comment]
Thank you very much, 
now I can build it

**Comment by Lastique** [comment]
See https://github.com/resiprocate/resiprocate/pull/250.
